src/Oracle/OracleService.cs

```diff
@@ -0,0 +1,485 @@
+        {
+            lock (_lock)
+            {
+                var task = PendingQueue.GetOrAdd(requestId, new OracleTask
+                {
+                    Id = requestId,
+                    Request = NativeContract.Oracle.GetRequest(snapshot, requestId),
+                    Signs = new ConcurrentDictionary<ECPoint, byte[]>(),
+                    BackupSigns = new ConcurrentDictionary<ECPoint, byte[]>()
+                });
+
+                if (responseTx != null)
+                {
+                    task.Tx = responseTx;
+                    var data = task.Tx.GetHashData();
+                    task.Signs.Where(p => !Crypto.VerifySignature(data, p.Value, p.Key)).ForEach(p => task.Signs.Remove(p.Key, out _));
+                }
+                if (backupTx != null)
+                {
+                    task.BackupTx = backupTx;
+                    var data = task.BackupTx.GetHashData();
+                    task.BackupSigns.Where(p => !Crypto.VerifySignature(data, p.Value, p.Key)).ForEach(p => task.BackupSigns.Remove(p.Key, out _));
+                    task.BackupSigns.TryAdd(oraclePub, backupSign);
+                }
+                if (task.Tx == null)
+                {
+                    task.Signs.TryAdd(oraclePub, sign);
+                    task.BackupSigns.TryAdd(oraclePub, sign);
+                    return;
+                }
+
+                if (Crypto.VerifySignature(task.Tx.GetHashData(), sign, oraclePub))
+                    task.Signs.TryAdd(oraclePub, sign);
+                else if (Crypto.VerifySignature(task.BackupTx.GetHashData(), sign, oraclePub))
+                    task.BackupSigns.TryAdd(oraclePub, sign);
+                else
+                    throw new RpcException(-100, "Invalid response transaction sign");
+
+                if (CheckTxSign(snapshot, task.Tx, task.Signs) || CheckTxSign(snapshot, task.BackupTx, task.BackupSigns))
+                {
+                    FinishedCache.TryAdd(requestId, new DateTime());
+                    PendingQueue.Remove(requestId, out _);
+                }
+            }
+        }
+
+        private byte[] Filter(string input, string filterArgs)
+        {
+            if (string.IsNullOrEmpty(filterArgs))
+                return NUtility.StrictUTF8.GetBytes(input);
+
+            NJObject beforeObject = NJObject.Parse(input);
+            NJArray afterObjects = new NJArray(beforeObject.SelectTokens(filterArgs, true));
+            return NUtility.StrictUTF8.GetBytes(afterObjects.ToString());
+        }
+
+        private bool CheckTxSign(StoreView snapshot, Transaction tx, ConcurrentDictionary<ECPoint, byte[]> Signs)
+        {
+            ECPoint[] nodes = NativeContract.Designation.GetDesignatedByRole(snapshot, Role.Oracle, snapshot.Height + 1);
+            int m = nodes.Length - (nodes.Length - 1) / 3;
```

Comment:
maybe let's call this as threshold for signatures, `neededThreshold` because the code is even iterating reducing it